Bug of using std::sync::Arc in Verus

<!-- 

Thanks for helping improve Verus!

In addition to describing the issue, please consider using the `--record` feature to produce a zip file (containing a copy of your crate, along with relevant metadata such as Verus command line arguments), to attach to the issue to help us reproduce it. See https://github.com/verus-lang/verus/blob/main/CONTRIBUTING.md for more details.

If you cannot (for any reason) attach your crate's source for reproducing, consider including a minimized example, along with your `verus --version`.  We have some support for automatic minimization of issues. See https://github.com/verus-lang/verus/tree/main/source/tools/minimizers for more details.

For feature requests, please use the Discussions page (https://github.com/verus-lang/verus/discussions).  We suggest adding a comment if you can find a relevant discussion, but feel free to open a new discussion in the "Feature Requests" category (https://github.com/verus-lang/verus/discussions/categories/feature-requests).

 -->

An example,

```
use vstd::prelude::*;
use std::sync::Arc;

verus!{

pub struct Obj{
    a : usize,
}
impl Obj{
    spec fn g(&self) -> usize {
        self.a
    }
}

fn test(x:Arc<Obj>){
    let ghost z = x.g();
}
```

For the function above, Rust reports an error

```
error[E0507]: cannot move out of an `Arc`
  --> ./test_arc.rs:19:19
   |
19 |     let ghost z = x.g();
   |                   ^ move occurs because value has type `test_arc::Obj`, which does not implement the `Copy` trait
   |
note: if `test_arc::Obj` implemented `Clone`, you could clone the value
  --> ./test_arc.rs:9:1
   |
 9 | pub struct Obj{
   | ^^^^^^^^^^^^^^ consider implementing `Clone` for this type
...
19 |     let ghost z = x.g();
   |                   - you could clone this value

error: aborting due to 1 previous error
```

--- a/source/rust_verify_test/tests/lifetime.rs
+++ b/source/rust_verify_test/tests/lifetime.rs
@@ -1564,3 +1564,80 @@ test_verify_one_file! {
         }
     } => Ok(())
 }
+
+test_verify_one_file! {
+    #[test] derefs_issue1971 verus_code! {
+        use std::sync::Arc;
+        use std::rc::Rc;
+        use vstd::*;
+
+        pub struct Obj{
+            a : usize,
+        }
+        impl Obj{
+            spec fn g(&self) -> usize {
+                self.a
+            }
+        }
+
+        fn test(x: Arc<Obj>){
+            let ghost z = x.g();
+        }
+
+        fn test2(x: Rc<Obj>){
+            let ghost z = x.g();
+        }
+
+        fn test3(x: Box<Obj>){
+            let ghost z = x.g();
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] derefs_lifetime verus_code! {
+        use std::sync::Arc;
+        use std::rc::Rc;
+        use vstd::*;
+
+        pub struct Obj{
+            a : usize,
+        }
+
+        fn consume<T>(t: T) { }
+
+        fn test(x: Arc<Obj>) {
+            let y: &Obj = &x;
+
+            consume(x);
+
+            proof {
+                let tracked z = y;
+            }
+        }
+    } => Err(err) => assert_rust_error_msg(err, "cannot move out of `x` because it is borrowed")
+}
+
+test_verify_one_file! {
+    #[test] derefs_lifetime2 verus_code! {
+        use std::sync::Arc;
+        use std::rc::Rc;
+        use vstd::*;
+
+        pub struct Obj{
+            a : usize,
+        }
+
+        fn consume<T>(t: T) { }
+
+        fn test(x: Arc<Obj>) {
+            let tracked y: &Obj = &x;
+
+            consume(x);
+
+            proof {
+                let tracked z = y;
+            }
+        }
+    } => Err(err) => assert_rust_error_msg(err, "cannot move out of `x` because it is borrowed")
+}

--- a/source/rustc_mir_build_additional_files/verus_expr.rs
+++ b/source/rustc_mir_build_additional_files/verus_expr.rs
@@ -226,7 +226,13 @@ pub(crate) fn apply_adjustment_post<'tcx>(
     spec: bool,
 ) -> rustc_middle::thir::ExprKind<'tcx> {
     match adjustment.kind {
-        Adjust::Deref(None) | Adjust::Borrow(AutoBorrow::Ref(_)) => {
+        Adjust::Deref(None | Some(_)) | Adjust::Borrow(AutoBorrow::Ref(_)) => {
+            // Adjust::Deref(None) -> implicit *
+            // Adjust::Borrow(AutoBorrow::Ref(_)) -> implicit &
+            // Adjust::Deref(Some(_)) -> This case means inserting a Deref::deref function.
+            //   In spec code that would usually be an error, except for some cases
+            //   like Arc or Rc where we ignore the deref in spec code.
+            //   In all those cases we also want to erase.
             maybe_erase_node(cx, expr, adjustment.target, kind, spec)
         }
         _ => kind,

